File: platform/backend/video/grok_imagine.py
"""Grok Imagine API integration for video generation.

Uses xAI's Grok Imagine API (OpenAI-compatible) for text-to-video generation.
API docs: https://x.ai/news/grok-imagine-api
"""

import os
import logging
from typing import Any
from openai import AsyncOpenAI

logger = logging.getLogger(__name__)

# Initialize client lazily
_client: AsyncOpenAI | None = None


def get_client() -> AsyncOpenAI:
    """Get or create the Grok Imagine API client."""
    global _client
    if _client is None:
        api_key = os.getenv("XAI_API_KEY")
        if not api_key:
            raise ValueError("XAI_API_KEY environment variable not set")
        _client = AsyncOpenAI(
            api_key=api_key,
            base_url="https://api.x.ai/v1",
        )
    return _client


async def generate_video(
    prompt: str,
    aspect_ratio: str = "16:9",
    duration: int = 5,
) -> dict[str, Any]:
    """Generate a video from a text prompt.

    Args:
        prompt: Description of the video to generate
        aspect_ratio: Video aspect ratio (16:9, 9:16, 1:1)
        duration: Video duration in seconds (max 8.7s)

    Returns:
        Dict with job_id, status, and url (when complete)
    """
    try:
        client = get_client()

        # Grok Imagine uses images.generate endpoint
        # Note: Size parameter is not supported by Grok API
        response = await client.images.generate(
            model="grok-2-image",
            prompt=prompt,
            n=1,
        )

        # For now, return image generation result
        # Video generation may use different endpoint when fully available
        if response.data:
            return {
                "status": "completed",
                "url": response.data[0].url,
                "revised_prompt": response.data[0].revised_prompt,
            }
        return {"status": "failed", "error": "No data in response"}

    except Exception as e:
        logger.error(f"Video generation failed: {e}")
        return {"status": "failed", "error": str(e)}


async def generate_story_video(
    world_name: str,
    world_premise: str,
    conversation_summary: str,
    characters: list[dict[str, str]],
    tone: str = "dramatic",
) -> dict[str, Any]:
    """Generate a short story video from world activity.

    Args:
        world_name: Name of the world
        world_premise: Core premise of the world
        conversation_summary: Summary of dweller conversation
        characters: List of character dicts with name and role
        tone: Video tone (dramatic, documentary, poetic, news)

    Returns:
        Dict with job_id, status, and url (when complete)
    """
    # Build cinematic prompt from story elements
    char_desc = ", ".join(
        f"{c['name']} ({c['role']})" for c in characters[:3]
    )

    tone_styles = {
        "dramatic": "cinematic, high contrast lighting, emotional close-ups",
        "documentary": "observational camera, natural lighting, intimate",
        "poetic": "ethereal, soft focus, metaphorical imagery",
        "news": "urgent, handheld camera, investigative",
    }
    style = tone_styles.get(tone, tone_styles["dramatic"])

    prompt = f"""Science fiction scene in {world_name}.
Setting: {world_premise[:200]}
Characters: {char_desc}
Scene: {conversation_summary[:300]}
Style: {style}
Mood: contemplative future, grounded sci-fi aesthetic"""

    result = await generate_video(prompt)

    # Add job tracking info
    if result.get("status") == "completed":
        import uuid
        result["job_id"] = str(uuid.uuid4())

    return result


async def generate_thumbnail(
    world_name: str,
    world_premise: str,
    style: str = "cinematic",
) -> dict[str, Any]:
    """Generate a thumbnail image for a world or story.

    Args:
        world_name: Name of the world
        world_premise: Core premise of the world
        style: Visual style for the thumbnail

    Returns:
        Dict with status and url
    """
    prompt = f"""Cinematic thumbnail for sci-fi world "{world_name}".
Setting: {world_premise[:300]}
Style: {style}, widescreen composition, dramatic lighting, no text"""

    return await generate_video(prompt)


def _derive_style_from_world(world_premise: str, year_setting: int) -> str:
    """Derive visual style keywords from world premise.

    Analyzes the premise to determine appropriate aesthetic.
    """
    premise_lower = world_premise.lower()

    # Check for common sci-fi aesthetics
    if any(w in premise_lower for w in ["cyber", "neon", "hacker", "digital", "virtual"]):
        return "cyberpunk aesthetic, neon accents, high contrast, tech wear"
    elif any(w in premise_lower for w in ["solar", "green", "ecology", "sustainable", "garden"]):
        return "solarpunk aesthetic, warm earth tones, organic shapes, hopeful"
    elif any(w in premise_lower for w in ["noir", "detective", "shadow", "rain", "dark"]):
        return "neo-noir aesthetic, dramatic shadows, muted colors, moody"
    elif any(w in premise_lower for w in ["desert", "sand", "wasteland", "post-apoc"]):
        return "post-apocalyptic aesthetic, dusty earth tones, rugged textures"
    elif any(w in premise_lower for w in ["space", "station", "orbit", "stellar"]):
        return "space opera aesthetic, sleek metallic surfaces, cosmic backdrop"
    elif any(w in premise_lower for w in ["bio", "genetic", "organic", "mutation"]):
        return "biopunk aesthetic, organic textures, unsettling beauty"
    elif year_setting > 2200:
        return "far future aesthetic, advanced materials, holographic accents"
    elif year_setting < 2050:
        return "near future aesthetic, recognizable technology, subtle enhancements"
    else:
        return "mid-century sci-fi aesthetic, retro-futuristic, clean lines"


async def generate_avatar(
    name: str,
    role: str,
    background: str,
    world_name: str,
    world_premise: str,
    year_setting: int,
) -> dict[str, Any]:
    """Generate a character avatar via xAI Grok.

    Creates a portrait that matches the world's aesthetic while
    representing the character's personality and role.

    Args:
        name: Character's name
        role: Character's role/occupation
        background: Character's background story
        world_name: Name of the world they inhabit
        world_premise: Core premise of the world (used for style)
        year_setting: Year the world is set in

    Returns:
        Dict with status, url, and prompt used
    """
    try:
        client = get_client()

        # Derive world style
        world_style = _derive_style_from_world(world_premise, year_setting)

        # Build character-focused prompt
        # Extract key personality traits from background
        background_snippet = background[:150] if background else ""

        prompt = f"""Portrait of {name}, a {role} in the year {year_setting}.

Character context: {background_snippet}

Setting: {world_name} - {world_premise[:100]}

Style: {world_style}

Requirements:
- Head and shoulders portrait, slight angle
- Thoughtful expression, character visible in eyes
- Clothing/styling appropriate to role and era
- Subtle environmental hints matching world
- Cinematic lighting, sharp focus on face
- No text or watermarks"""

        logger.info(f"Generating avatar for {name} in {world_name}")

        response = await client.images.generate(
            model="grok-2-image",
            prompt=prompt,
            n=1,
        )

        if response.data:
            return {
                "status": "completed",
                "url": response.data[0].url,
                "prompt": prompt,
                "revised_prompt": response.data[0].revised_prompt,
            }
        return {"status": "failed", "error": "No data in response", "prompt": prompt}

    except Exception as e:
        logger.error(f"Avatar generation failed for {name}: {e}")
        return {"status": "failed", "error": str(e), "prompt": ""}

File: platform/lib/video/grok-imagine.ts
/**
 * Grok Imagine API Integration
 *
 * Uses xAI's Grok Imagine API for video generation.
 * Max 8.7s videos at 480p/720p with native audio.
 *
 * API is compatible with OpenAI SDK.
 */

import OpenAI from 'openai'

// Initialize xAI client with OpenAI-compatible SDK
const xai = new OpenAI({
  apiKey: process.env.XAI_API_KEY,
  baseURL: 'https://api.x.ai/v1',
})

export interface VideoGenerationRequest {
  prompt: string
  aspectRatio?: '16:9' | '9:16' | '1:1'
  duration?: number // max 8.7 seconds
  resolution?: '480p' | '720p'
  imageUrl?: string // For image-to-video
}

export interface VideoGenerationResult {
  jobId: string
  status: 'pending' | 'processing' | 'complete' | 'failed'
  videoUrl?: string
  thumbnailUrl?: string
  error?: string
}

/**
 * Generate a video from text prompt
 */
export async function generateVideo(
  request: VideoGenerationRequest
): Promise<VideoGenerationResult> {
  try {
    // Use the images endpoint for video generation
    // Note: This is placeholder - actual API format may differ
    const response = await xai.images.generate({
      model: 'grok-2-image', // or grok-2-video when available
      prompt: request.prompt,
      n: 1,
      size: request.aspectRatio === '9:16' ? '1024x1792' : '1792x1024',
    })

    // For now, return a placeholder since video API isn't fully documented
    return {
      jobId: `job_${Date.now()}`,
      status: 'pending',
    }
  } catch (error) {
    console.error('Video generation error:', error)
    return {
      jobId: `job_${Date.now()}`,
      status: 'failed',
      error: error instanceof Error ? error.message : 'Unknown error',
    }
  }
}

/**
 * Check the status of a video generation job
 */
export async function checkVideoStatus(jobId: string): Promise<VideoGenerationResult> {
  // TODO: Implement job status checking when API is available
  return {
    jobId,
    status: 'pending',
  }
}

/**
 * Generate a story video from dweller conversation
 */
export async function generateStoryVideo(options: {
  worldName: string
  worldPremise: string
  conversationSummary: string
  characters: { name: string; role: string }[]
  tone?: 'dramatic' | 'contemplative' | 'hopeful' | 'tense'
}): Promise<VideoGenerationResult> {
  const { worldName, worldPremise, conversationSummary, characters, tone = 'dramatic' } = options

  // Craft a cinematic prompt
  const characterList = characters.map((c) => `${c.name} (${c.role})`).join(', ')

  const prompt = `Cinematic sci-fi scene. ${worldPremise}. ${conversationSummary}.
Characters: ${characterList}.
Style: ${tone}, professional cinematography, atmospheric lighting,
futuristic setting, no text overlays.
Camera: slow push in, shallow depth of field.`

  return generateVideo({
    prompt,
    aspectRatio: '16:9',
    duration: 8,
    resolution: '720p',
  })
}

/**
 * Generate a thumbnail for a story
 */
export async function generateThumbnail(options: {
  worldPremise: string
  storyTitle: string
}): Promise<string | null> {
  try {
    const response = await xai.images.generate({
      model: 'grok-2-image',
      prompt: `Cinematic movie poster style. ${options.worldPremise}.
Title: "${options.storyTitle}".
Style: dramatic lighting, sci-fi atmosphere, no text.`,
      n: 1,
      size: '1792x1024',
    })

    return response.data?.[0]?.url ?? null
  } catch (error) {
    console.error('Thumbnail generation error:', error)
    return null
  }
}
